### PR TITLE
Update Debian packaging to use the new configuration directories

### PR DESCRIPTION
**Pull request by mamarley**
This updates the Debian packaging to use the new configuration directories introduced by
https://github.com/tvheadend/tvheadend/pull/1535 and https://github.com/tvheadend/tvheadend/pull/1538.  Instead of being in /home/hts/.hts/tvheadend, the configuration will now be stored in /var/lib/tvheadend by default, which is consistent with how many other "server"-style Debian packages handle their home directories and configuration.

It also fixes the logic in config.c:config_get_dir() to properly support the forking mode of operation.  Since config_get_dir() is executed before forking, the uid will always be 0 at this point. Instead, use the uid of the user to which we will fork if a fork will occur.

### CODE REVIEW

**Comment by Flole998** [comment]
Just to be sure: This is a non-breaking change as it will only apply to new user accounts created during Installation?

File: debian/tvheadend.postinst [hunk 1]
```diff
@@ -10,16 +10,15 @@ configure)
 
    if ! getent passwd $HTS_USER >/dev/null; then
         echo >&2 "Creating user: $HTS_USER..."
-        adduser --quiet --system --group --shell /bin/bash $HTS_USER
+        adduser --quiet --system --group --home /var/lib/tvheadend --shell /bin/bash $HTS_USER
```
[the remark this answers is not in the document]
Comment (mamarley) [reply debian/tvheadend.postinst]: My testing confirms that; I have removed the shell.

**Comment by mamarley** [comment]
> Just to be sure: This is a non-breaking change as it will only apply to new user accounts created during Installation?

Correct, existing installations will be untouched both as far as the user and the TVH configuration itself.

**Comment by Flole998** [comment]
@oliv3r Do you see anything wrong with these changes or does it look like a good idea to you? If it's good I'll merge it.

**Comment by oliv3r** [comment]
@Flole998 i'll look asap

File: src/config.c [hunk 2]
```diff
@@ -1693,14 +1693,16 @@ config_check ( void )
 
 static int config_newcfg = 0;
 
-static char *config_get_dir ( void )
```
Comment (oliv3r) [inline src/config.c:1696]: When I was testing this (in a docker container mind you) I didn't have any issue with this. I don't think this is a bad idea of course, I didn't look into this too deeply!

Regardless, this should be its own commit, not connected to the debian packaging :)

File: src/config.c [hunk 3]
```diff
@@ -1693,14 +1693,16 @@ config_check ( void )
...
+static char *config_get_dir ( uid_t uid )
 {
-  char hts_home[PATH_MAX + sizeof("/.hts/tvheadend")]; /* Must be largest of the 3 config strings! */
```
Comment (oliv3r) [inline src/config.c:1698]: Why is this needed?

hts_home is either something (usually $HOME) + `/.hts/tvheadend` OR `/var/lib/tvheadend`, which fits into PATH_MAX just fine.

File: src/config.c [hunk 4]
```diff
@@ -1721,7 +1723,7 @@ static char *config_get_dir ( void )
 
       if ((readlink(hts_home, hts_home_link, sizeof(hts_home_link)) == -1) ||
           (stat(hts_home_link, &st) == -1)) {
-        tvherror(LS_CONFIG, ".hts/tvheadend is inaccessable: %s", strerror(errno));
+        tvherror(LS_CONFIG, ".hts/tvheadend is inaccessible: %s", strerror(errno));
```
Comment (oliv3r) [inline src/config.c:1726]: oops :)

File: src/main.c [hunk 5]
```diff
@@ -1113,7 +1113,7 @@ main(int argc, char **argv)
   signal(SIGPIPE, handle_sigpipe); // will be redundant later
   signal(SIGILL, handle_sigill);   // see handler..
 
-  /* Set priviledges */
+  /* Set privileges */
```
Comment (oliv3r) [inline src/main.c:1116]: oops again! my dyslexia combined with English being my non-native language doesn't help.

I recommend to put the two spelling fixes in one commit, the UID fix in a second commit. Personally, I'm okay with having this under the debian package MR, but that's upto @Flole998

File: debian/tvheadend.postinst [hunk 6]
```diff
@@ -10,16 +10,15 @@ configure)
 
    if ! getent passwd $HTS_USER >/dev/null; then
         echo >&2 "Creating user: $HTS_USER..."
-        adduser --quiet --system --group --shell /bin/bash $HTS_USER
+        adduser --quiet --system --group --home /var/lib/tvheadend $HTS_USER
```
Comment (oliv3r) [inline debian/tvheadend.postinst:13]: what about migration? Maybe add a check 'if old, inform user of new setup?

File: debian/tvheadend.postinst [hunk 7]
```diff
@@ -10,16 +10,15 @@ configure)
...
    fi
 
    HTS_HOME=`getent passwd $HTS_USER | cut -d':' -f6`
 
-   mkdir -p "${HTS_HOME}/.hts/tvheadend"
-   chown ${HTS_USER}:${HTS_USER} "${HTS_HOME}/.hts"
-   chown ${HTS_USER}:${HTS_USER} "${HTS_HOME}/.hts/tvheadend"
+   mkdir -p "${HTS_HOME}/recordings"
```
Comment (oliv3r) [inline debian/tvheadend.postinst:18]: `install` is better here

`install -d -m 0775 -g 'tvheadend' -o 'tvheadend' '/var/lib/tvheadend/recordings'` is what I'm using in my container atm

File: debian/tvheadend.postinst [hunk 8]
```diff
@@ -10,16 +10,15 @@ configure)
...
+   chown ${HTS_USER}:${HTS_USER} -R "${HTS_HOME}"
```
Comment (oliv3r) [inline debian/tvheadend.postinst:19]: not needed with install :p

File: debian/tvheadend.postinst [hunk 9]
```diff
@@ -10,16 +10,15 @@ configure)
...
 
-   HTS_SUPERUSERCONF="${HTS_HOME}/.hts/tvheadend/superuser"
+   HTS_SUPERUSERCONF="${HTS_HOME}/superuser"
```
Comment (oliv3r) [inline debian/tvheadend.postinst:21]: no idea what this is for, so can't judge this.

**Comment by Saentist** [comment]
How current configuration will be handled by new package?
What about auto-backup during updating?

File: debian/tvheadend.postinst (hunk 6 above)
Comment (Flole998) [reply debian/tvheadend.postinst:13]: This won't get executed on existing installs, so there is no migration. As it isn't a benefit on existing Installations I'd assume that it can stay as-is.

File: debian/tvheadend.postinst (hunk 7 above)
Comment (Flole998) [reply debian/tvheadend.postinst]: Some limited OSs might not have the install command, so I'd leave that as is, you never know where people try to install it ;)

File: src/config.c (hunk 2 above)
Comment (mamarley) [reply src/config.c:1696]: Were you using the forking mode of operation when you tested?  The problem I encountered is that the Debian package forks (starts as root and then drops privileges), so getuid() still returns 0 when the config directory is checked.  In the case that forking is not used, the passed uid will be -1 and getuid() will then be used as you originally wrote.

File: src/config.c (hunk 4 above)
Comment (mamarley) [reply src/config.c:1726]: No problem, it happens.

File: src/main.c (hunk 5 above)
Comment (mamarley) [reply src/main.c:1116]: I don't think this one is yours; I think it was there before.

File: debian/tvheadend.postinst (hunk 6 above)
Comment (mamarley) [reply debian/tvheadend.postinst:13]: This particular bit of code only executes on new installs, not upgrades.  Detecting when such an upgrade occurred would be potentially problematic, so I didn't go off into those weeds.  Plus, the code you added already has a warning when the old configuration directory is in use.

File: debian/tvheadend.postinst (hunk 7 above)
Comment (mamarley) [reply debian/tvheadend.postinst]: Cool, I didn't know that was a thing.

**Comment by mamarley** [comment]
> How current configuration will be handled by new package? What about auto-backup during updating?

Current installations will not be touched in any way, shape, or form, so no backup is necessary.  This change affects only new installations.

File: debian/tvheadend.postinst (hunk 7 above)
Comment (mamarley) [reply debian/tvheadend.postinst]: @Flole998 , "install" is a part of coreutils and has been since at least coreutils 7.  All supported Debian and Ubuntu versions have at least 8.21 and this code only ever gets executed on Debianoid systems, so I think we are good there.

**Comment by mamarley** [comment]
Thanks for the input, everyone.  I think I have addressed all the suggestions.

File: src/config.c (hunk 3 above)
Comment (oliv3r) [reply src/config.c:1698]: Yeah, but we never do $HOME/var/lib/tvheadend :) the goal is, to have `hts_home` be large enough. Ideally, the var stays at `PATH_MAX`, and everything else gets reduced to fit, but that's far more complex :p this is much lazyer ..easier :)

So we need room to store $HOME + whatever; OR `/etc/tvheadend` | `/var/lib/tvheadend`. The later two, easily fit within `PATH_MAX`.

File: src/main.c (hunk 5 above)
Comment (oliv3r) [reply src/main.c:1116]: It's a typo I make very often though :)

File: debian/tvheadend.postinst (hunk 6 above)
Comment (oliv3r) [reply debian/tvheadend.postinst:13]: so true :) sorry for the noise

File: debian/tvheadend.postinst (hunk 7 above)
Comment (oliv3r) [reply debian/tvheadend.postinst]: i think `install` is a posix requirement ...

Also, this will only ever be run on debian will it not, which will guaranteed have `install` :p

**Comment by oliv3r** [comment]
> How current configuration will be handled by new package? What about auto-backup during updating?

"like always", My changes keep existing setups 'as is' (with a warning).

File: src/config.c (hunk 2 above)
Comment (oliv3r) [reply src/config.c:1696]: Ah! No! I did look into `geteuid()` (afaik) but your solutions is cleaner anyway. I was testing the docker container, which will run tvheadend as <user> anyhow.

Thanks for the catch!

File: src/config.c (hunk 3 above)
Comment (mamarley) [reply src/config.c:1698]: Good point, I had failed to consider that.  I changed it back.

**Comment by Flole998** [comment]
I will let the build run through and merge it tomorrow when I get near a computer :)
